ssz-rs/src/de.rs

```diff
@@ -93,26 +102,41 @@ where
         return Ok(vec![])
     }
 
-    let data_pointer = u32::deserialize(&encoding[..BYTES_PER_LENGTH_OFFSET])?;
-    let data_pointer = data_pointer as usize;
-    if encoding.len() < data_pointer {
+    let offsets_len = encoding.get(..BYTES_PER_LENGTH_OFFSET).ok_or({
+        DeserializeError::ExpectedFurtherInput {
+            provided: encoding.len(),
+            expected: BYTES_PER_LENGTH_OFFSET,
+        }
+    })?;
+    let offsets_len = u32::deserialize(offsets_len)?;
+    let offsets_len = offsets_len as usize;
+    if encoding.len() < offsets_len {
         return Err(DeserializeError::ExpectedFurtherInput {
             provided: encoding.len(),
-            expected: data_pointer,
+            expected: offsets_len,
         })
     }
+    if offsets_len % BYTES_PER_LENGTH_OFFSET != 0 {
+        return Err(DeserializeError::InvalidOffsetsLength(offsets_len))
+    }
 
-    let offsets = &mut encoding[..data_pointer]
+    let offsets = &mut encoding[..offsets_len]
         .chunks_exact(BYTES_PER_LENGTH_OFFSET)
         .map(|chunk| u32::deserialize(chunk).map(|offset| offset as usize))
         .collect::<Result<Vec<usize>, DeserializeError>>()?;
     offsets.push(encoding.len());
 
-    let element_count = data_pointer / BYTES_PER_LENGTH_OFFSET;
+    let element_count = offsets_len / BYTES_PER_LENGTH_OFFSET;
     let mut result = Vec::with_capacity(element_count);
     for span in offsets.windows(2) {
```

Comment:
also worth noting that I believe as far as this code can be reached, there will also be at least 2 elements in `offsets` so it is fine to call `windows(2)` here and not worry about skipping/missing some data

Reply:
There's technically a case with 1 element in `offsets`, but it doesn't look like a problem:

If `offsets_len == 0` then `offsets == vec![encoding.len()]`. In this case we return an empty `Vec` (both `len == 0` and `cap == 0`). This seems reasonable though, because this function is for variable sized composite types like `List` that are allowed to be empty.